pkg/controller/network/internetgateway/controller.go

```diff
@@ -97,125 +115,140 @@ type external struct {
-	// their ID.
 	if meta.GetExternalName(cr) == "" {
-		return managed.ExternalObservation{ResourceExists: false}, nil
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
 	}
 
-	req := e.client.DescribeInternetGatewaysRequest(&awsec2.DescribeInternetGatewaysInput{
+	response, err := e.client.DescribeInternetGatewaysRequest(&awsec2.DescribeInternetGatewaysInput{
 		InternetGatewayIds: []string{meta.GetExternalName(cr)},
-	})
-
-	response, err := req.Send(ctx)
-	if ec2.IsInternetGatewayNotFoundErr(err) {
-		return managed.ExternalObservation{ResourceExists: false}, nil
-	}
+	}).Send(ctx)
 	if err != nil {
-		return managed.ExternalObservation{}, errors.Wrap(err, errDescribe)
+		return managed.ExternalObservation{}, errors.Wrapf(resource.Ignore(ec2.IsInternetGatewayNotFoundErr, err), errDescribe)
 	}
 
 	// in a successful response, there should be one and only one object
 	if len(response.InternetGateways) != 1 {
-		return managed.ExternalObservation{}, errors.New(errMultipleItems)
+		return managed.ExternalObservation{}, errors.Errorf(errNotSingleItem)
 	}
 
-	observed := response.InternetGateways[0]
+	cr.SetConditions(runtimev1alpha1.Available())
 
-	// if non of the attachments are currently in progress, then the IG is available
-	isAvailable := true
-	for _, a := range observed.Attachments {
-		if a.State == awsec2.AttachmentStatusAttaching || a.State == awsec2.AttachmentStatusDetaching {
-			isAvailable = false
-			break
-		}
-	}
+	observed := response.InternetGateways[0]
 
-	if isAvailable {
-		cr.SetConditions(runtimev1alpha1.Available())
-	}
+	cr.Status.AtProvider = ec2.GenerateIGObservation(observed)
 
-	cr.UpdateExternalStatus(observed)
+	isUptoDate := ec2.IsIgUpToDate(cr.Spec.ForProvider, observed)
 
 	return managed.ExternalObservation{
-		ResourceExists:    true,
-		ConnectionDetails: managed.ConnectionDetails{},
+		ResourceExists:   true,
+		ResourceUpToDate: isUptoDate,
 	}, nil
 }
 
 func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
-	cr, ok := mgd.(*v1alpha3.InternetGateway)
+	cr, ok := mgd.(*v1beta1.InternetGateway)
 	if !ok {
 		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
 	}
 
-	req := e.client.CreateInternetGatewayRequest(&awsec2.CreateInternetGatewayInput{})
+	cr.Status.SetConditions(runtimev1alpha1.Creating())
 
-	rsp, err := req.Send(ctx)
+	ig, err := e.client.CreateInternetGatewayRequest(&awsec2.CreateInternetGatewayInput{}).Send(ctx)
 	if err != nil {
 		return managed.ExternalCreation{}, errors.Wrap(err, errCreate)
 	}
 
-	meta.SetExternalName(cr, aws.StringValue(rsp.InternetGateway.InternetGatewayId))
-	if err := e.kube.Update(ctx, cr); err != nil {
-		return managed.ExternalCreation{}, errors.Wrap(err, errPersistExternalName)
+	meta.SetExternalName(cr, aws.StringValue(ig.InternetGateway.InternetGatewayId))
+
+	return managed.ExternalCreation{}, errors.Wrap(e.kube.Update(ctx, cr), errSpecUpdate)
```

Comment:
Here and elsewhere in this PR, it seems like we have a call order of:

1. Set conditions. (Modifies status)
1. Set external name. (Modifies metadata)
1. Call `e.kube.Update()`

This is all followed by a call to `e.kube.Status().Update()` made by the `managed.Reconciler`.

Will calling `Update()` before we call `Status().Update()` undo the status conditions that we set? I can never remember whether `Update()` overwrites status or not.

Reply:
> Will calling Update() before we call Status().Update() undo the status conditions that we set?

Yes. The operation that undoes the rest of the resource is not `Update` but the call itself actually. What it does when you call either `Status().Update()` or `Update()` is that it issues the respective payload (either only `status` or `metadata+spec`) and then deep copies the resulting object into what you have sent. So, the local object is the most up-to-date version that exists in the api-server, which means the payload that you didn't send is overwritten by whatever is in api-server.

We have synced with @sahil-lakhwani about this and it seems calling `Status().Update()` after the `SetConditions` makes sense because that way if the `CreateInternetGatewayRequest` call fails, the condition will reflect that the error arose during creation.